relayer/chains/bitcoin/provider.go

```diff
@@ -0,0 +1,1437 @@
+	key := "RB" + sn.String()
+	p.logger.Info("stored rollback message key", zap.String("key", key))
+	err = p.db.Put([]byte(key), data, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to store rollback message data: %v", err)
+	}
+	return relayMessage, nil
+}
+
+func (p *Provider) verifyBridgeTx(tx *TxSearchRes, messageInfo *multisig.XCallMessage) (string, *big.Int, bool, error) {
+	verified := false
+	isMatchAmount := false
+	runeId := ""
+	runeAmount := big.NewInt(0)
+	amount := big.NewInt(0)
+	amount.SetBytes(messageInfo.Amount)
+
+	for i, out := range tx.Tx.TxOut {
+		if !bytes.Equal(out.PkScript, p.multisigAddrScript) {
+			continue
+		}
+
+		if messageInfo.TokenAddress == p.cfg.BTCToken {
+			verified = true
+			btcAmount := big.NewInt(out.Value)
+			if amount.Cmp(btcAmount) == 0 {
+				isMatchAmount = true
+				break
+			} else {
+				messageInfo.Amount = btcAmount.Bytes()
+			}
+		} else {
+			runes, err := p.GetUTXORuneBalance(p.cfg.UniSatURL, tx.Tx.TxHash().String(), i)
+			if err != nil {
+				return "", nil, false, err
+			}
+
+			if len(runes.Data) == 0 {
+				continue
+			}
+
+			for _, runeOut := range runes.Data {
+				runeTokenBal, ok := big.NewInt(0).SetString(runeOut.Amount, 10)
+				if !ok {
+					return "", nil, false, fmt.Errorf("rune amount out invalid")
+				}
+
+				if amount.Cmp(runeTokenBal) == 0 && runeOut.RuneId == messageInfo.TokenAddress {
+					runeId = runeOut.RuneId
+					runeAmount = runeTokenBal
+					// runes tx is not verified if amount is not match
+					verified = true
+					isMatchAmount = true
+					break
+				}
+			}
+		}
+	}
+	if !verified {
+		return "", nil, false, fmt.Errorf("failed to verify transaction %v", tx.Tx.TxHash().String())
+	}
+	return runeId, runeAmount, isMatchAmount, nil
+}
+
+func (p *Provider) getMessagesFromTxList(resultTxList []*TxSearchRes) ([]*relayTypes.BlockInfo, error) {
+	var messages []*relayTypes.BlockInfo
+	for _, resultTx := range resultTxList {
+		msg, err := p.parseMessageFromTx(resultTx)
+		if err != nil {
+			p.logger.Error("Failed to parse message from tx", zap.Error(err))
+			continue
+		}
+		if msg == nil {
+			continue
+		}
+
+		msg.MessageHeight = resultTx.Height
+		p.logger.Info("Detected eventlog",
+			zap.Uint64("height", msg.MessageHeight),
+			zap.String("target_network", msg.Dst),
+			zap.Uint64("sn", msg.Sn.Uint64()),
+			zap.String("event_type", msg.EventType),
+		)
+		messages = append(messages, &relayTypes.BlockInfo{
+			Height:   resultTx.Height,
+			Messages: []*relayTypes.Message{msg},
+		})
+	}
+	return messages, nil
+}
+
+func (p *Provider) getNumOfPipelines(diff int) int {
+	if diff <= runtime.NumCPU() {
+		return diff
+	}
+	return runtime.NumCPU() / 2
+}
+
+func (p *Provider) runBlockQuery(ctx context.Context, blockInfoChan chan *relayTypes.BlockInfo, fromHeight, toHeight uint64) uint64 {
+	done := make(chan bool)
+	defer close(done)
+
+	heightStream := p.getHeightStream(done, fromHeight, toHeight)
+
+	diff := int(toHeight-fromHeight) / 2
```

Comment:
Similar implementation on other chains have been reverted to use simpler single threaded executions to avoid hitting rate limit in RPCs when numerous blocks are missing. So, it would be better to refactor this as well